addons/purchase_stock/models/purchase.py

```diff
@@ -270,6 +270,27 @@ def create(self, values):
 
     @api.multi
     def write(self, values):
+        propagated_date_field = False
+        propagated_changes_dict = {}
+
+        if values.get('date_planned'):
+            propagated_date_field = 'date_planned'
+
+        for line in self:
+            if line.move_dest_ids and line.propagate_date:
+                if 'date_expected' in propagated_changes_dict:
+                    propagated_changes_dict.pop('date_expected')
+                if propagated_date_field:
+                    current_date = line.date_planned
```

Comment:
@hsh-odoo why all this cruft here? you don't need propagated_date_field and propagated_changes_dict. you can simplify the code a lot here. also, the second par of the code always update date_expected whatever propagate_date_minimum_delta, so i'm not sure i can trust the tests you wrote?

``` python
for line in self:
    if values.get('date_planned'):
       current_date = line.date_planned
       new_date = fields.Datetime.from_string(values['date_planned'])
       delta_days = (new_date - current_date).total_seconds() / 86400
       if abs(delta_days) < line.propagate_date_minimum_delta:
           continue
       moves_to_update = line.move_ids.filtered(lambda m. m state not in ('done', 'cancel'))
       if not moves_to_update:
           moves_to_update = line.move_dest_ids.filtered(lambda m. m state not in ('done', 'cancel'))
       for move in moves_to_update:
           move.date_expected = move.date_expected + relativedelta.relativedelta(days=delta_days)
result = super(PurchaseOrderLine, self).write(values)
if 'product_qty' in values:
    self.filtered(lambda l: l.order_id.state == 'purchase')._create_or_update_picking()
return result
```

Reply:
you also need to updte _find_candidate and probably _run_buy